### DISCUSSION TITLE
Troubles with Javascript sort

### DISCUSSION THREAD

**Discussion by pviltard** [apache/couchdb#3991] [Q&A]
Hello,

I have troubles with the sort Javascript with documents.
I use CouchDb in windows 2012R2 :
{
  "couchdb": "Welcome",
  "version": "3.2.0",
  "git_sha": "efb409bba",
  "uuid": "cd3e089da5227286e25d8d0155ba59f7",
  "features": [
    "access-ready",
    "partitioned",
    "pluggable-storage-engines",
    "reshard",
    "scheduler"
  ],
  "vendor": {
    "name": "The Apache Software Foundation"
  }
}

I found several articles on the Stack Overflow site, but I’m stuck.

My documents look like this example :

{
  "_id": "DOC:10",
  "_rev": "1-bd10f4e4000d236b3fe4ffb5852f0204",
  "sequence": 10,
  "composedflow": 3262656,
  "userid": "2010000758915",
  "composedflowuserid": "1XXXXL.P.VALID_CTX_V5.PCTXB5-B_055_R3_100114_00001428_94_----------.00091614.000028563",
  "date": "2009-09-07T00:00:00",
  "zone": [
    {
      "label": "Code modèle",
      "value": "ADV013",
      "number": 9
    },
    {
      "label": "Index Générique B",
      "value": "null",
      "number": 41
    },
    {
      "label": "Index Générique A",
      "value": "null",
      "number": 40
    },
    {
      "label": "Numéro d'inscription au répertoire (N° INSEE ou Sécu)",
      "value": "null",
      "number": 39
    },
    {
      "label": "Identifiant Applicatif National (allocataire / Affilié)",
      "value": "null",
      "number": 38
    }
  ]
}

I would like to create a view by sorting the "zone" area objects by the "number" field.
Nothing could be simpler.

First view :

function (doc) {
  if(doc.zone) {
    var list = doc.zone;
       emit( doc._id, list); 
  }
}

I do get an object array in the value with fauxton 👍 

Second view :
function (doc) {
  if(doc.zone) {
    var list = doc.zone.sort(function(a,b) { return a.number-b.number });
       emit( doc._id, list); 
  }
}

"No Documents Found" by Fauxton in less than a second. There are 12,000 Json Documents 👎 

My third view :
function (doc) {
  if(doc.zone) {
    var temp = [
       {
      "label": "Code modèle",
      "value": "ADV013",
      "number": 9
    },
    {
      "label": "Index Générique B",
      "value": "null",
      "number": 41
    },
    {
      "label": "Index Générique A",
      "value": "null",
      "number": 40
    },
    {
      "label": "Numéro d'inscription au répertoire (N° INSEE ou Sécu)",
      "value": "null",
      "number": 39
    },
    {
      "label": "Identifiant Applicatif National (allocataire / Affilié)",
      "value": "null",
      "number": 38
    }
               ];
    var list = temp.sort(function(a,b) { return a.number-b.number });
       emit( doc._id, list); 
  }
}

Good but static data 👍 👎 

Thank you so much for your help.
Pascal

**Comment by big-r81** [comment]
Hi Pascal,

doesn't know why "Array.sort" don't work atm, but maybe you can use this view function, to get a sorted list of numbers:

```javascript
function (doc) {
  if(doc.zone) {
    doc.zone.forEach(zone_el => {
      emit([doc._id, zone_el.number], null);
    });
  }
}
```

You iterate through the zone array and put the numbers into the key so that you get sorting for "free".

**Comment by big-r81** [comment]
Hi Pascal,

testing and thinking about your problem...

If you use the sort-function on the document array "zone", it would in-place sort the array. This is not allowed in a view function, because this would change the whole document (the "zone" element)! What you need to do instead is to copy the array, sort the copy and emit the result. For example you can use this map function:

```javascript
function (doc) {
  if(doc.zone) {
    let zone_sorted = [...doc.zone].sort((a, b) => a.number - b.number);
    emit(doc._id, zone_sorted);
  }
}
```

The answer is a little bit hidden [here](https://docs.couchdb.org/en/stable/ddocs/ddocs.html#map-functions):

> Also keep in mind that each document is sealed to prevent the situation where one map function changes document state and another receives a modified version.

**Comment by pviltard** [comment]
Hi,

The Array is well sorted by the Big-r81 solution.
I would like to make a change in the assumptions:

function (doc) {
if(doc.zone) {
let zone_sorted = [...doc.zone].sort((a,b) => a.number - b.number);
zone_sorted.forEach( function(item){
emit( doc._id, item); }
);
}
}

And why emited element in the zone_sorted are not sorted ?

thank you for your help.
Pascal

**Comment by big-r81** [comment]
Hi,

but why would you do that? Then you can use the first provided approach.

Maybe @janl can help with the js part?
Why is a sorted array not in order with ForEach? Isn't it guaranteed if running in parallel?

Ronny

**Reply by pviltard** [reply]
Hello Ronny,

(I have some problems with the GitHub interface and health problems).
This is a migration of a large Oracle partitioned table that will be put in CouchDb.
It is an "octopus" that represents 2/3 of the volume of the database, which points on all the biggest tables of the application. These are the information areas.

In the example above, DOC:10 (Document Number 10) has 5 information areas stored in an array (zone).
I performed the bulk loading of the information areas using the utl_http package of Oracle and PL/JSON (Open Source on Github).

I am designing the views for Java developers and I thought that CouchDb’s javascript will give me a lot of flexibility. CouchDb uses spiderMonkey. (https://guide.couchdb.org/draft/source.html)

I have trouble understanding the unsorted result

if(doc.zone) {
let zone_sorted = [...doc.zone].sort((a,b) => a.number - b.number);
zone_sorted.forEach( function(item){
emit( doc._id, item); }
);
}

and the notation [...doc.zone] is now understandable by https://oprearocks.medium.com/what-do-the-three-dots-mean-in-javascript-bc5749439c9a.

I ask these questions because I have other more complicated views, but are they achievable?
I was surprised by the second view with the result : No Document

Thank you for your valuable assistance.
Merci de votre précieuse aide.

Pascal

**Comment by big-r81** [comment]
Hi Pascal,

> function (doc) { if(doc.zone) { let zone_sorted = [...doc.zone].sort((a,b) => a.number - b.number); zone_sorted.forEach( function(item){ emit( doc._id, item); } ); } }

I can't tell you at the moment why this isn't working (needs further investigation), but you have other possibilities that I mentioned which are working.

The question you have to ask yourself is how you want to make the data of a view available to the user. I can't answer this for you, but I might be able to help you create a map reduce function.

Please read the documentation for [design documents](https://docs.couchdb.org/en/stable/ddocs/index.html), it explains a lot.

Afterwards, tell your problems and we try to create a view that shows you your data as you (hopefully) want it.

Ronny

**Reply by pviltard** [reply]
Hello Ronny,

I was able to speak with one of the Java developers, the solution seems to strongly interest them.

`function (doc) {
  if(doc.zone) {
    let zone_sorted = [...doc.zone].sort((a, b) => a.number - b.number);
    emit(doc._id, zone_sorted);
  }
}`

Thank you for your support and time.

Pascal
